File: uilib/src/main/java/dev/entao/ui/list/swipe/MoveHandler.java
package dev.entao.ui.list.swipe;

import android.os.Handler;
import android.os.Message;
import android.view.View;

/**
 * Created by yet on 2015/10/26.
 */
public class MoveHandler extends Handler {
    private final static int DURATION = 100;
    private final static int DURATION_STEP = 10;
    int stepX = 0;
    int fromX = 0;
    int toX = 0;
    View view;

    public MoveHandler(View view, int to) {
        this.view = view;
        this.fromX = view.getScrollX();
        this.toX = to;
        stepX = (int) ((toX - fromX) * DURATION_STEP * 1.0 / DURATION);
    }

    @Override
    public void handleMessage(Message msg) {
        if (Math.abs(toX - fromX) <= Math.max(Math.abs(stepX), 10)) {
            view.scrollTo(toX, 0);
        } else {
            fromX += stepX;
            view.scrollTo(fromX, 0);
            this.sendEmptyMessageDelayed(0, DURATION_STEP);
        }
    }
}

File: uilib/src/main/java/dev/entao/ui/list/MoveHandler.kt
package dev.entao.ui.list

import android.os.Handler
import android.os.Message
import android.view.View

/**
 * Created by yet on 2015/10/26.
 */
class MoveHandler(internal var view: View, to: Int) : Handler() {
    var stepX = 0
    var fromX = 0
    var toX = 0

    init {
        this.fromX = view.scrollX
        this.toX = to
        stepX = ((toX - fromX).toDouble() * DURATION_STEP.toDouble() * 1.0 / DURATION).toInt()
    }

    override fun handleMessage(msg: Message) {
        if (Math.abs(toX - fromX) <= Math.max(Math.abs(stepX), 10)) {
            view.scrollTo(toX, 0)
        } else {
            fromX += stepX
            view.scrollTo(fromX, 0)
            this.sendEmptyMessageDelayed(0, DURATION_STEP.toLong())
        }
    }

    companion object {
        private const val DURATION = 100
        private const val DURATION_STEP = 10
    }
}